File: 5-longest-palindromic-substring/5-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        bool dp[1001][1001];
        int sLen = s.size();
        string ans = s.substr(0,1); 
        memset(dp, false, sizeof(dp));
        
        for (int i = 0; i < sLen; i++) {
            dp[i][i] = true;
            if (i) dp[i][i - 1] = true;
        }
        
        for (int len = 2; len <= sLen; len++) {
            for (int i = 0; i <= sLen - len; i++) {
                if (s[i] == s[i + len - 1] && dp[i + 1][i + len - 2]) {
                    if (ans.size() < len) {
                        ans = s.substr(i, len);
                    }
                    dp[i][i + len - 1] = true;
                }
            }
        }
        
        return ans;
    }
};

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        
        sLen = len(s)
        answer = s[0:1]
        
        dp = [[False] * sLen for row in range(sLen)]
        
        for i in range(sLen): 
            dp[i][i] = True
            if i > 0:
                dp[i][i - 1] = True
            
        for gap in range(1, sLen):
            for i in range(sLen):
                
                if i + gap < sLen and s[i] == s[i + gap]:
                    if dp[i + 1][i + gap - 1]:
                        dp[i][i + gap] = True
                        answer = s[i:i + gap + 1]

        return answer